File: tools/e2e-cli/src/tests/ping.ts
import assert from 'assert';

import { TestContext, TestDefinition } from '../types';

async function testPing(ctx: TestContext): Promise<void> {
  const type = ctx.adapter.type;

  // Always run from a prepared project: its .npmrc carries the fresh token and
  // shields the test from whatever the developer's global ~/.npmrc has for
  // this registry (a stale token there would turn into a spurious E401).
  const { tempFolder } = await ctx.adapter.prepareProject(
    `verdaccio-ping-${ctx.runId}`,
    '1.0.0',
    ctx.registryUrl,
    ctx.port,
    ctx.token
  );
  const cwd = tempFolder;
  if (type === 'yarn-modern' && ctx.adapter.importPlugin) {
    await ctx.adapter.importPlugin(tempFolder, 'npm-ping');
  }

  const jsonFlag = type === 'yarn-modern' ? [] : ['--json'];
  const resp = await ctx.adapter.exec(
    { cwd },
    'ping',
    ...jsonFlag,
    ...ctx.adapter.registryArg(ctx.registryUrl)
  );

  if (type === 'yarn-modern') {
    // yarn npm ping succeeding (exit code 0) is sufficient proof the registry is reachable
    assert.ok(true, 'ping succeeded');
  } else {
    const parsedBody = JSON.parse(resp.stdout);
    assert.strictEqual(
      parsedBody.registry,
      ctx.registryUrl + '/',
      `Expected registry "${ctx.registryUrl}/" but got "${parsedBody.registry}"`
    );
  }
}

export const pingTest: TestDefinition = {
  name: 'ping',
  requires: ['ping'],
  run: testPing,
};

File: tools/e2e-cli/src/e2e_tests/ping.rs
use super::prelude::*;

pub(crate) fn test_ping(ctx: &mut TestContext<'_>) -> Result<()> {
    let cwd = if ctx.adapter.kind == AdapterType::YarnModern {
        let temp = prepare_project(
            ctx,
            &format!("verdaccio-ping-{}", ctx.run_id),
            "1.0.0",
            BTreeMap::new(),
            BTreeMap::new(),
        )?;
        import_plugin(ctx, &temp, "npm-ping")?;
        Some(temp)
    } else {
        None
    };
    let mut args = vec!["ping".into()];
    if ctx.adapter.kind != AdapterType::YarnModern {
        args.push("--json".into());
    }
    args.extend(registry_arg(&ctx.adapter, &ctx.registry_url));
    let resp = exec_adapter(ctx, cwd.as_deref(), args)?;
    if ctx.adapter.kind != AdapterType::YarnModern {
        let parsed: Value = serde_json::from_str(&resp.stdout)?;
        let expected_registry = format!("{}/", ctx.registry_url);
        assert_eq_json(
            parsed.get("registry"),
            expected_registry.as_str(),
            "Registry mismatch",
        )?;
    }
    Ok(())
}
